api/rds.proto

```diff
@@ -47,6 +47,10 @@ message WeightedCluster {
     // an upstream cluster is determined by its weight. The sum of weights
     // across all entries in the clusters array must add up to 100.
     google.protobuf.UInt32Value weight = 2;
+    // Optional endpoint metadata match criteria. Only endpoints in the upstream
+    // cluster with metadata matching that set in metadata_match will be
+    // considered. The filter name should be specified as "envoy.lb".
+    Metadata metadata_match = 3;
```

Comment:
I still don't think this makes any sense at all. Not to mention, it will severely muddy the config and make it hard to understand the system. @9len @mccv the example you showed for traffic splitting (dynamic clustering) was just a YAML spec. If you look at the actual config, it goes like this
https://lyft.github.io/envoy/docs/configuration/http_conn_man/route_config/traffic_splitting.html#traffic-splitting-across-multiple-upstreams

```json
{
  "route_config": {
    "virtual_hosts": [
      {
        "name": "helloworld",
        "domains": ["*"],
        "routes": [
          {
            "prefix": "/",
            "weighted_clusters": {
              "runtime_key_prefix" : "routing.traffic_split.helloworld",
              "clusters" : [
                { "name" : "helloworld_v1", "weight" : 33 },
                { "name" : "helloworld_v2", "weight" : 33 },
                { "name" : "helloworld_v3", "weight" : 34 }
              ]
            }
          }
        ]
      }
    ]
  }
}
```

Now, if you want to use a single cluster and split based on endpoint metadata dynamically, you would specify the same cluster name for all three clusters, but use different weights. This would require some changes to existing code as well, because the runtime paths associated with each weighted-cluster is assumed to be unique (`routing.traffic_split.helloworld.helloworld_v1`). The code checks these runtime paths for overrides before applying the values specified in the config file. Doing this metadata based thing would require removing runtime support for weight adjustment (or serialize the endpoint info into runtime paths, with some specific semantics on how paths are constructed).